lib/sqlalchemy/ext/mypy/apply.py

```diff
@@ -242,13 +266,14 @@ def _add_additional_orm_attributes(
                     kind=ARG_NAMED_OPT,
                 )
             )
+
         add_method_to_class(api, cls, "__init__", arguments, NoneTyp())
 
-    if "__table__" not in info.names and cls_metadata.has_table:
+    if "__table__" not in info.names and util.get_has_table(info):
         _apply_placeholder_attr_to_class(
             api, cls, "sqlalchemy.sql.schema.Table", "__table__"
         )
-    if cls_metadata.is_mapped:
+    if not is_base:
```

Comment:
**mike bayer** (zzzeek) wrote:

not every class in an ORM hierarchy has __mapper__ directly on the class, there's __abstract__ and mixin classes that would not have this directly; does this accommodate for that?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2962